feat: openapi-implementation

dasarathg68 opened the pull request:
<h1> Description </h1>
The PR filed addresses issue #12 and presents OpenAPI specification for the Node.js backend.
swagger-autogen package is used to automatically parse the routes and identify paths, parameters and also descriptions.
The generated document is served using swagger-ui-express npm package on the /api-docs endpoint.
Further, the controller functions are grouped using by using tags. (Teams, Members)
Here's how it looks.
<img width="1512" alt="Screenshot 2024-03-30 at 12 07 56 PM" src="https://github.com/globe-and-citizen/cnc-portal/assets/31665486/5ff8d9c3-27d3-4160-ad0a-1f3509fc8359">
<img width="1512" alt="Screenshot 2024-03-30 at 12 08 31 PM" src="https://github.com/globe-and-citizen/cnc-portal/assets/31665486/efc7c816-48f3-4eec-802f-f4753a6c09ea">

<h1> Type of Change </h1>
Non breaking change that implements OpenAPI for description of the APIs.

hermannleboss commented on backend/package.json:
```diff
@@ -3,6 +3,11 @@
   "version": "1.0.0",
   "description": "",
   "main": "src/index.ts",
+  "nodemonConfig": {
+    "ignore": [
+      "src/utils/swagger-output.json"
+    ]
+  },
```
Why this is necessary

dasarathg68 replied:
During dev we use nodemon to watch for changes right? 
It creates an infinite loop as every time the swagger-output is generated, it assumes it to be a change and restarts the server over and over again. 
Thats why this has to be present

hermannleboss replied:
OK

hermannleboss commented on backend/src/controllers/memberController.ts:
```diff
@@ -50,10 +56,13 @@ const updateMember = async (req: Request, res: Response) => {
 
 //delete members
 const deleteMembers = async (req: Request, res: Response) => {
+  /* 
+  #swagger.tags = ['Members']
+  */
   const { id } = req.params;
 
   try {
-    const member = await prisma.member.delete({
+    const member = await prisma.member.deleteMany({
```
Why this change
And even if it's an optimization or a correction, it has nothing to do with swagger and open API.
So it shouldn't be in this PR

dasarathg68 replied:
This change is about swagger.
swagger assumes anything in the pattern .delete() to be api endpoints.
So prisma.member.delete() would be considered to be an api endpoint which is not the case.

hermannleboss commented on backend/package.json:
Look at this if it can be helpfull
https://www.npmjs.com/package/swagger-jsdoc

dasarathg68 replied:
Thanks but I already took a look at that package and it cannot be used since we're not generating openapi specification. We're using swagger-autogen library which uses minimal directives to generate the specification automatically and swagger-ui-express to render the specification.

hermannleboss commented on backend/src/utils/swagger-output.json:
If it's generated everytime we run the server do we need to have it inside git ?

dasarathg68 replied:
We need the file to be present. It is a requirement of the swagger-autogen library.
If we don't add to git then we have to create it manually each time.
Atleast an empty file named swagger-output.json with just {} should be present

hermannleboss reviewed (changes requested):
Good starting point. 

Look at this API,
https://petstore.swagger.io/


The api have to explain herself. 

- [x] Every input should have a type, 
- [x] we should know if it's optional or not. 
- [ ] And we should know the type of the output 
- [ ] and his status

Try to implement theses requirements

dasarathg68 commented:
> Good starting point.
> 
> Look at this API, https://petstore.swagger.io/
> 
> The api have to explain herself.
> 
> * [x]  Every input should have a type,
> * [x]  we should know if it's optional or not.
> * [x]  And we should know the type of the output
> * [x]  and his status
> 
> Try to implement theses requirements

@hermannleboss i have updated/reverted the code to work with the manual version and i have addressed the requirements. Please check the documentation at localhost:3000/api-docs after running the backend.
